src/Mockly.Generator/CodeGeneration/ContainingHierarchyBuilder.cs

```diff
@@ -0,0 +1,96 @@
+﻿using System.Text;
+
+using Mockly.Generator.CodeGeneration.Extensions;
+
+namespace Mockly.Generator.CodeGeneration;
+
+public class ContainingHierarchyBuilder
+{
+    private int _hierarchyLevelCount = 0;
+    private readonly Dictionary<int, string> _indentationCache = new();
+    private readonly StringBuilder _builder;
+
+    public int CurrentIndentationLevel { get; private set; }
+
+    public ContainingHierarchyBuilder(StringBuilder builder)
+    {
+        _builder = builder;
+        CurrentIndentationLevel = 0;
+    }
+
+    public ContainingHierarchyBuilder AddContainingHierarchy(IEnumerable<INamedTypeSymbol> containingClassSymbols, string containingNamespace)
+    {
+        AddContainingNamespace(containingNamespace);
+
+        foreach (var containingClassSymbol in containingClassSymbols)
+        {
+            AddContainingClass(containingClassSymbol);
+        }
+
+        return this;
+    }
+
+    private ContainingHierarchyBuilder AddContainingNamespace(string containingNamespace)
+    {
+        _builder.AppendLine($"namespace {containingNamespace}");
+        _builder.AppendLine("{");
+        CurrentIndentationLevel += Constants.IndentationPerLevel;
+        return this;
+    }
+
+    private ContainingHierarchyBuilder AddContainingClass(INamedTypeSymbol containingClassSymbol)
+    {
+        _hierarchyLevelCount++;
+
+        var indentation = GetOrCreateIndentationLevel(CurrentIndentationLevel);
+        var containingClassAccessibility = containingClassSymbol.GetAccessibilityString();
+
+        var containingTypeStatic = containingClassSymbol.IsStatic ? "static " : string.Empty;
+
+        var containingTypeKind = containingClassSymbol.TypeKind switch
+        {
+            TypeKind.Class => "class",
+            TypeKind.Struct => "struct",
+            _ => throw new InvalidOperationException("Only classes and structs can be used as containing types."),
+        };
+
+        var containingClassName = containingClassSymbol.Name;
+
+        _builder.AppendLine($"{indentation}{containingClassAccessibility} {containingTypeStatic}partial {containingTypeKind} {containingClassName}");
```

Comment:
Might be something for later, but do we want to add a `sealed` modifier option?

Reply:
Sealed only needs to be on 1 part of the definition of the partial containing structure, so this would be up to the user.